### PR TITLE
New style dropout function

### PR DESCRIPTION
**Pull request by bonprosoft**
This PR implements new-style version of `F.dropout`.
It supports double backward for Chainer v3.
It is related to #3147 .

### CODE REVIEW

**Comment by mitmul** [comment]
Jenkins, test this please.

**Comment by mitmul** [comment]
Jenkins, test this please

File: chainer/functions/noise/dropout.py [hunk 1]
```diff
@@ -45,7 +44,21 @@ def forward(self, x):
...
+
+    def __init__(self, mask):
+        self.mask = mask
+
+    def forward(self, inputs):
+        y = inputs[0] * self.mask
+        return y,
+
+    def backward(self, indexes, gy):
+        return DropoutGrad(self.mask).apply((gy[0],))
```
Comment (mitmul) [inline chainer/functions/noise/dropout.py]: `(gy[0],)` can be just `gy`.

File: chainer/functions/noise/dropout.py [hunk 2]
```diff
@@ -81,5 +94,6 @@ def dropout(x, ratio=.5, **kwargs):
     argument.assert_kwargs_empty(kwargs)
 
     if configuration.config.train:
-        return Dropout(ratio)(x)
+        y, = Dropout(ratio).apply((x,))
+        return y
```
Comment (mitmul) [inline chainer/functions/noise/dropout.py]: replace this line with `return Dropout(ratio).apply((x,))[0]` and remove L:97 is OK here, I think

File: tests/chainer_tests/functions_tests/noise_tests/test_dropout.py [hunk 3]
```diff
@@ -43,21 +45,19 @@ def check_forward(self, x_data):
...
-        x = chainer.Variable(x_data)
-        y = functions.dropout(x, self.ratio)
-        creator = y.creator
-        y.grad = y_grad
-        y.backward()
+        dropout = functions.Dropout(self.ratio)
+
+        def f(x):
+            x, = dropout.apply((x,))
+            return x
```
Comment (mitmul) [inline tests/chainer_tests/functions_tests/noise_tests/test_dropout.py]: `return dropout.apply((x,))[0]` can reduce the number of lines.

**Comment by bonprosoft** [comment]
@mitmul Thank you for your review. I fixed them. Please check again.

**Comment by mitmul** [comment]
Jenkins, test this please

**Comment by mitmul** [comment]
LGTM

**Pull request merged**